---
title: "XD Analysis"
author: "Jess Diaz"
date: "`r format(Sys.time(), '%a %d %b')`"
output:
  rmdformats::readthedown
---

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# load libraries
library(qiime2R)
library(microbiome)
library(dplyr)
library(ggplot2)
library(vegan)
library(decontam)
library(phyloseq)
library(rmdformats)
source("HighstatLibV10.R")
library(ANCOMBC)
library(tibble)
library(pairwiseAdonis)

# set graphing options
theme_set(theme_light())
sc = scale_color_manual(values = c("black", "darksalmon", "darkred"))
```

```{r functions, include=FALSE}
# calculates values for stats bars for plots
data.summary <- function(x) {
  m <- mean(x)
  ymin <- m-sd(x)
  ymax <- m+sd(x)
  return(c(y=m, ymin=ymin, ymax=ymax))
}
```

This is the analysis of the Xenopus tadpole thyroxine experiment. There are X samples plus 12 kit controls. These samples are spread across 3 groups: Thyroxine, Delayed Thyroxine, and Control.  

```{r create phyloseq object, include = FALSE}
# create master phyloseq
XD <-qza_to_phyloseq(
  features="table-filt-bytaxa.qza",
  tree="rooted-tree.qza",
  taxonomy="taxonomy.qza",
  metadata = "XD_qiime_metadata.txt"
)

# sum contrasts for categorical variables
sample_data(XD)$tank.id <- C(factor(sample_data(XD)$tank.id), sum)
sample_data(XD)$treatment <- C(factor(sample_data(XD)$treatment), sum)
```

# Data Cleanup

## Summary

```{r data summary, echo = FALSE}
# data summaries
summarize_phyloseq(XD)
```
## Decontam

First I will run decontam using the 12 kit control samples as the negative controls.

```{r decontam, echo = FALSE, message = FALSE}
# make new variable where controls are "TRUE"
sample_data(XD)$is.neg <- sample_data(XD)$sample.type == "Blank"

# run decontam
contamdf.prevalence <- isContaminant(XD, method="prevalence", neg="is.neg", threshold=0.5)

# identify which taxa are contaminants
table(contamdf.prevalence$contaminant)
contams <- which(contamdf.prevalence$contaminant)

# list contaminants
tax <- as(tax_table(XD), "matrix")
tax[contams,]
```
I will also make a plot that shows prevalence in the negatives and in true samples, to check how they are getting classified.  
```{r decontam plot, echo = FALSE, message = FALSE}
# presence-absence phyloseq for controls and samples
ps.pa <- transform_sample_counts(XD, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$sample.type == "Blank", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$sample.type == "Gut", ps.pa)

# make dataframe for prevalence
df.pa <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                    contaminant=contamdf.prevalence$contaminant)

# plot prevalence
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) +
  geom_point() +
  xlab("Prevalence (blanks)") +
  ylab("Prevalence (samples)") +
  ggtitle("Threshold = 0.5") +
  geom_abline(slope = 1, intercept = 0)
```

24 taxa were identified as contaminants using a threshold of 0.5. Based on the  plot above it seems reasonable to me to classify these as contaminants. Now I just want to make sure that I am not removing too many samples by removing these contaminants. This table shows % of reads removed from that sample after removing contaminants.  

```{r remove contaminants, echo = FALSE, message = FALSE}
# remove taxa from phyloseq
XD.decontam <- prune_taxa(!contamdf.prevalence$contaminant, XD)

# set up df with readcounts and % removed
readcounts <- data.frame(matrix(ncol = 0, nrow = 291))
readcounts$pre <- as.data.frame(sample_sums(XD))
readcounts$post <- sample_sums(XD.decontam)
readcounts$per.removed <- ((readcounts$pre - readcounts$post)/readcounts$pre) *100

# print percent of reads in each sample that were removed
print(readcounts$per.removed)

# print otu table for contaminants
as.data.frame(otu_table(XD)) %>% slice(contams)
```
I am happy with how this looks, the couple samples with higher % of reads removed had low read counts anyway. Last, I need to remove frogs from the analysis (stage >= 66).

```{r phyloseq cleanup, include = FALSE}
# remove samples that were allowed to metamorphose
XD.clean <- subset_samples(XD.decontam, age.days <= 65)
```

# Exploring Metadata

## Initial plots

```{r meta, echo = FALSE}
# isolate metadata for ggplot
meta <- meta(XD.clean)
```

First I want to make the age/stage plot, colored by treatment group.
```{r Age vs Stage, echo = FALSE}
# age and stage plot, line connecting means
meta %>% 
  ggplot(aes(x = age.days, y = stage, color = treatment)) +
  geom_point(size = 3, shape = 16) +
  stat_summary(geom = "line", fun = mean, linewidth = 1.5) +
  labs(x = "Age (days)", y = "Developmental Stage") +
  sc
```

Then I want to look at the relationship between body length and mass:

```{r body length and mass, echo = FALSE}
# mass vs length plot
meta %>% 
  ggplot(aes(x = length.mm, y = mass.g)) +
  geom_point(size = 3, aes(color = treatment)) + sc +
  labs(x = "Body length (mm)", y = "Mass (g)")
```
There don't seem to be any red points in the upper part of the graph, meaning the thyroxine group was smaller than the other groups. Maybe thyroxine stunted growth by accelerating development? Is is worth (or possible) to explore this with a linear model?  

## Body Condition variable 

I want to make a new Body Condition measure, using the residuals of a linear regression between mass and body length (see Bancila et al 2010). First I want to transform the variables to give them a more linear relationship. Doing the log10 of both seems to work, and this is also what was done in Denoel et al 2002).  

```{r log body length and mass, echo = FALSE}
# plot log-transformed mass vs length
meta %>% 
  ggplot(aes(x = log10(length.mm), y = log10(mass.g))) +
  geom_point(size = 3, aes(color = treatment)) + sc +
  labs(x = "Log10(Body length (mm))", y = "Log10(Mass (g))")
```

To find the index I will first make the log variables, then run the model and save the residuals as a body condition variable. (I am using this as an opportunity to also add the Y/N thyroxine exposure category to be used for differential abundance)  

```{r body condition model, echo = FALSE, message = FALSE}
# add log-transformed length and mass
meta <- meta %>% 
  mutate(length.mm.log = log10(length.mm)) %>% 
  mutate(mass.g.log = log10(mass.g))

# run model for log mass (response) and log length (explanatory)
BCI.model <- lm(mass.g.log ~ length.mm.log, data = meta)
summary(BCI.model)

# plot the model prediction
plot(mass.g.log ~ length.mm.log, data = meta)
x <- seq(0.38, 1.4, length.out = 10)
lines(x, -3.60608 + 2.82665 * x, col = "red")

# add residuals to metadata as body condition measure
# add variable saying whether tadpole was ever exposed to thyroxine
meta <- meta %>% 
  mutate(body.condition = BCI.model$residuals) %>% 
  mutate(exposureYN = ifelse(days.exposed >0, "YES", "NO"))

# integrate variables into phyloseq object
meta_to_add <- meta %>%
  select(body.condition, exposureYN)
meta_to_add.ps <- sample_data(meta_to_add)
XD.complete <- merge_phyloseq(XD.clean, meta_to_add.ps)
meta.complete <- meta(XD.complete)
```

Now I can look at other variables in relation to body condition. For example, how does body condition change with age?  
```{r BCI and age, echo = FALSE}
meta.complete %>% 
  ggplot(aes(x = age.days, y = mass.g, shape = treatment)) +
  geom_point(size = 3, aes(color = treatment)) + sc +
  labs(x = "Age (days)", y = "Mass (g)")

meta.complete %>% 
  ggplot(aes(x = age.days, y = body.condition, shape=treatment)) +
  geom_point(size = 3, aes(color = treatment)) + sc +
  labs(x = "Age (days)", y = "Body condition")
```

Mass increases as the groups age but body condition doesn't, likely because they are growing proportionately in length as the get bigger in mass.

```{r BCI and treatment, echo = FALSE}
meta.complete %>% 
  ggplot(aes(x = treatment, y = body.condition)) +
  geom_boxplot(size = 1, aes(color = treatment)) + sc +
  labs(x = "Treatment", y = "Body Condition")
```
Treatment doesn't seem to impact body condition. What about the number of days exposed to thyroxine?  

```{r BCI and days exposed, echo = FALSE}
meta.complete %>% 
  ggplot(aes(x = days.exposed, y = body.condition)) +
  geom_point(size = 3, aes(color = treatment)) + sc +
  labs(x = "Days exposed to thyroxine", y = "Body Condition")
```

Body condition also doesn't seem to be changing depending on how long the animals have been exposed to thyroxine.  

## Statistics: Which facors influence growth?

Next, I want to assess whether thyroxine exposure influences the effect of age on determining tadpole developmental stage.
[analysis to come soon]

# Rarefaction

Now I can look at the read counts to see how many reads are in each sample.
```{r read counts, echo = FALSE}
sample_sums(XD.complete)
```
Based on the read counts, I will rarefy at 3638 to retain all data points except the five that had fewer reads. Blanks should be automatically removed with the rarefaction.

```{r rarefy, include=FALSE}
XD.rare <- rarefy_even_depth(XD.complete, sample.size=3638, rngseed=14, verbose=TRUE) # rarefy
meta.rare <- meta(XD.rare)
```

# Alpha diversity

## Plots

I can look at the initial alpha diversity plots to look for differences. These look pretty similar.  
```{r alpha all, echo = FALSE}
plot_richness(XD.rare, x="treatment")
```
More zoomed in richness plots for each metadata variable:  

```{r alpha setup, include = FALSE}
# save alpha diversity calculations as alpha dataframe
alpha <- microbiome::alpha(XD.rare)

# save phyloseq metadata as separate object, and add shannon and observed richness
meta.rare <- meta.rare %>%
  mutate(shannon = alpha$diversity_shannon) %>%
  mutate(observed = alpha$observed)
```

```{r alpha plots, echo = FALSE}
# by group
meta.rare %>%
  ggplot(aes(x=treatment, y=shannon, shape=treatment)) + sc +
  geom_point(aes(color = treatment), position = position_dodge(width = 0.1), size = 3, alpha = 0.5) +
  stat_summary(fun.data = data.summary, geom = "errorbar", linewidth = 1, width = 0.5, na.rm = TRUE, aes(color = treatment)) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by age
meta.rare %>%
  ggplot(aes(x=age.days, y=shannon, shape=treatment)) + sc +
  geom_point(aes(color = treatment), size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by stage
meta.rare %>%
  ggplot(aes(x=stage, y=shannon, shape=treatment)) + sc +
  geom_point(aes(color = treatment), size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by days exposed
meta.rare %>%
  ggplot(aes(x=days.exposed, y=shannon, shape=treatment)) + sc +
  geom_point(aes(color = treatment), size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by mass
meta.rare %>%
  ggplot(aes(x=mass.g, y=shannon, shape=treatment)) + sc +
  geom_point(aes(color = treatment), size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")
```
No clear patterns here. Maybe diversity is increasing a little with age but overall there don't seem to be any visual relationships explaining variation in alpha diversity.  

## Statistics: What factors determine variation in alpha diversity?

Response variable: shannon
Explanatory variables: dna.ng.ul, tank.id, mass.g, length.mm, stage, treatment, age.days, readcount, exposureYN, body.condition.
Based on the plots I will remove length since it covaries with mass, and will log transform DNA and mass.  
I will also scale the age and stage variables so that I can directly compare their effects in the model.

```{r pairs plot, warning = FALSE, echo = FALSE}
meta.rare$exposureYN <- C(factor(meta.rare$exposureYN), sum)
alpha.pairs <- meta.rare %>% 
  select(-date.collected, -dissector, -sample.type, -days.exposed, -is.neg, -observed) 

# Make pairs plot
pairs(alpha.pairs, lower.panel=panel.smooth, diag.panel=panel.hist, upper.panel=panel.cor)

# Calculate VIFs
corvif(subset(alpha.pairs, select=-c(shannon, treatment, tank.id, exposureYN)))
corvif(subset(alpha.pairs, select=-c(shannon, treatment, tank.id, exposureYN, length.mm)))

# New histograms
hist(log(alpha.pairs$dna.ng.ul))
hist(log(alpha.pairs$mass.g))

# Make transformations
alpha.pairs.transformed <- alpha.pairs %>% 
  mutate(dna.ng.ul.LOG = log(dna.ng.ul)) %>% 
  mutate(mass.g.LOG = log(mass.g)) %>% 
  select(-mass.g, -dna.ng.ul)

# New plots
pairs(alpha.pairs.transformed, lower.panel=panel.smooth, diag.panel=panel.hist,
      upper.panel=panel.cor)

# Scale age and stage variables
alpha.pairs.scaled <- alpha.pairs.transformed %>% 
  mutate(age.days.S = scale(age.days)) %>% 
  mutate(stage.S = scale(stage)) %>% 
  select(-age.days, -stage, -length.mm)
```

Now to fit the model: I am chooosing not to use tank as a random effect because I think it had some problems when I did that for my biostats project.

```{r initial model}
model <- lm(shannon ~ tank.id + dna.ng.ul.LOG + mass.g.LOG + treatment + 
              body.condition + exposureYN + age.days.S + stage.S, 
             data = alpha.pairs.scaled)
summary(model)
```
I believe there is a problem including both tank and treatment because the treatments were given to specific tanks. I will remove the treatment variable and try again.

```{r initial model redo}
model <- lm(shannon ~ tank.id + dna.ng.ul.LOG + mass.g.LOG + 
              body.condition + exposureYN + age.days.S + stage.S, 
             data = alpha.pairs.scaled)
summary(model)
```

Now I can compare this to other model variations. Namely:  
1. Shannon ~ Age + Stage  
2. Shannon ~ Age + Stage + Treatment  
3. Shannon ~ Age + Stage + Tank ID  
4. Shannon ~ Age + Stage + Mass 
5. Shannon ~ Age + Stage + DNA  
6. Shannon ~ Age + Stage + Body condition  
7. Shannon ~ Age + Stage + ExposureYN  
8. Shannon ~ Age * Stage + Tank ID + DNA + Mass + Body condition + ExposureYN  

```{r model comparison}
model1 <- lm(shannon ~ age.days.S + stage.S, 
             data = alpha.pairs.scaled)
model2 <- lm(shannon ~ age.days.S + stage.S + treatment, 
             data = alpha.pairs.scaled)
model3 <- lm(shannon ~ age.days.S + stage.S + tank.id, 
             data = alpha.pairs.scaled)
model4 <- lm(shannon ~ age.days.S + stage.S + mass.g.LOG, 
             data = alpha.pairs.scaled)
model5 <- lm(shannon ~ age.days.S + stage.S + dna.ng.ul.LOG, 
             data = alpha.pairs.scaled)
model6 <- lm(shannon ~ age.days.S + stage.S + body.condition, 
             data = alpha.pairs.scaled)
model7 <- lm(shannon ~ age.days.S + stage.S + exposureYN, 
             data = alpha.pairs.scaled)
model8 <- lm(shannon ~ tank.id + dna.ng.ul.LOG + mass.g.LOG + 
              body.condition + exposureYN + age.days.S * stage.S, 
             data = alpha.pairs.scaled)

# Find the AICs
AIC(model1)
AIC(model2)
AIC(model3)
AIC(model4)
AIC(model5)
AIC(model6)
AIC(model7)
AIC(model8)
AIC(model)
```
The best model seems to be model 5, which makes sense given the very high significance of DNA concentration in the first model. I will try one more version of model 5 with an interaction term between age and stage.

```{r model 5}
model5.2 <- lm(shannon ~ age.days.S * stage.S + dna.ng.ul.LOG, 
             data = alpha.pairs.scaled)
AIC(model5.2)
```
It looks like this was about the same so I will exclude the interaction term. Now model validation:  

```{r model validation}
# Residuals vs fitted values
plot(residuals(model5, type="response") ~ model5$fitted.values)
abline(0,0,col="gray")

# Residuals vs Age
plot(residuals(model5, type="response") ~ model5$model$age.days.S)
abline(0,0,col="gray")

# Residuals vs Stage
plot(residuals(model5, type="response") ~ model5$model$stage.S)
abline(0,0,col="gray")

# Residuals vs body length
plot(residuals(model5, type="response") ~ model5$model$dna.ng.ul.LOG)
abline(0,0,col="gray")
```

The residuals look fine to me. Can proceed with model interpretation.

```{r model results, echo = FALSE, warnings = FALSE, message = FALSE}
summary(model5)

# test significance of terms
car::Anova(model5)
```
DNA concentration was by far the biggest predictor of shannon diversity in these tadpoles, which limits our interpretation of what may be happening biologically. Both age and stage were also significant predictors, with stage explaining slightly more variation in shannon diversity. However, DNA concentration is likely not interpretable in this experiment because these concentrations reflect ALL DNA that was isolated from the gut samples, including both microbiome and host DNA as well as non-microbial/non-host DNA like viruses, fungi, etc. Because DNA concentration may not be relevant as an interpretable biological variable, we will remove it as an explanatory variable and proceed with new models using the other variables.

Trying more versions of the model without DNA included.  

10.1. Shannon ~ Age + Stage + Mass + Body condition + ExposureYN + Tank ID  
10.2. Shannon ~ Age + Stage + Mass + Body condition + Tank ID  
10.3. Shannon ~ Age + Stage + Mass + Tank ID  
10.4. Shannon ~ Age * Stage + Mass + Tank ID  
10.5. Shannon ~ Age * Stage  

```{r new models}

model10.1 <- lm(shannon ~ age.days.S + stage.S + mass.g.LOG + body.condition + exposureYN + tank.id, 
             data = alpha.pairs.scaled)
model10.2 <- lm(shannon ~ age.days.S + stage.S + mass.g.LOG + body.condition + tank.id, 
             data = alpha.pairs.scaled)
model10.3 <- lm(shannon ~ age.days.S + stage.S + mass.g.LOG + tank.id, 
             data = alpha.pairs.scaled)
model10.4 <- lm(shannon ~ age.days.S * stage.S + mass.g.LOG + tank.id, 
             data = alpha.pairs.scaled)
model10.5 <- lm(shannon ~ age.days.S * stage.S, 
             data = alpha.pairs.scaled)

# Find the AICs
AIC(model10.1)
AIC(model10.2)
AIC(model10.3)
AIC(model10.4)
AIC(model10.5)
```
Overall, the models with DNA included are a better fit, as expected by the previous results. When DNA is not included, models 1, 6, 7, 10.5, and 4 are the next best fits. To recap what those were:  
1. Shannon ~ Age + Stage  
6. Shannon ~ Age + Stage + Body condition  
7. Shannon ~ Age + Stage + ExposureYN 
10.5. Shannon ~ Age * Stage 
4. Shannon ~ Age + Stage + Mass 

Two of these are just age and stage alone (with the best one having no interaction term), and the other three are age and stage plus one other variable. I am going to dig deeper into the age/stage alone model, to see the relative weights of each variable in that model.  

```{r model10.5 validation}
# Residuals vs fitted values
plot(residuals(model1, type="response") ~ model1$fitted.values)
abline(0,0,col="gray")

# Residuals vs Age
plot(residuals(model1, type="response") ~ model1$model$age.days.S)
abline(0,0,col="gray")

# Residuals vs Stage
plot(residuals(model1, type="response") ~ model1$model$stage.S)
abline(0,0,col="gray")
```

```{r model 10.5 results, echo = FALSE, warnings = FALSE, message = FALSE}
summary(model10.5)

# test significance of terms
car::Anova(model10.5)
```
Interestingly, age is appearing as super significant while stage is non-significant. This aligns with the rest of the data that tends to show age having a greater effect than stage.

# Beta diversity

## Plots

```{r beta div prep}
BrayDist <- distance(XD.rare, method = "bray")
set.seed(14)
BrayOrd <- ordinate(XD.rare, "NMDS", distance = BrayDist)
BrayPlot <- plot_ordination(XD.rare, BrayOrd, justDF = TRUE)
```

Plot of all samples:
```{r main beta div, echo = FALSE}
BrayPlot %>% 
  ggplot(aes(x = NMDS1, y = NMDS2, shape=treatment)) +
  geom_point(size = 3, aes(color = treatment)) + sc
```
Now I will color the plots by age and stage:  

```{r colored bray curtis, echo = FALSE}
# colored by age
BrayPlot %>%
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  geom_point(size = 3, aes(color = age.days)) +
  scale_colour_gradient(low = "lightblue", high = "navy") +
  labs(color = "Age (days)")

# colored by stage
BrayPlot %>%
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  geom_point(size = 3, aes(color = stage)) +
  scale_colour_gradient(low = "lightblue", high = "navy") +
  labs(color = "Developmental Stage")
```

Age seems to have clearer separation over time compared to stage, but they both move from left to right. This likely explains the clustering of the Delayed Thyroxine group since they were all older tadpoles.  

I can also color by exposure to thyroxine.  

```{r thyroxine beta div, echo = FALSE}
BrayPlot %>%
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  geom_point(size = 3, aes(color = days.exposed)) +
  scale_colour_gradient(low = "lightblue", high = "navy") +
  labs(color = "Days Exposed")
```
Thyroxine exposure itself doesn't seem to be driving this since you have tadpoles with 0 days exposed all over the graph. Age/stage definitely seem to be the major drivers here.

## Statistics: PERMANOVA

Using a PERMANOVA I should be able to see how various variables influence beta diversity. My variables of interest are treatment, age, and stage. I also want to include tank ID as a random effect, but cannot do that here so I will try to include it as a fixed effect.

```{r permanovas, echo = FALSE}
adonis2(BrayDist ~ treatment + age.days + stage + tank.id, data = meta.rare)
```

Everything is very significant except stage, but the F-stat for age is WAY higher than any of the others.

The other stat to add is to see how the treatment groups cluster using a pairwise adonis.
```{r pairwise Adonis, echo = FALSE}
# pairwise by treatment
pairwise.adonis(BrayDist, meta.rare$treatment)
```

It looks like the control and thyroxine groups may have significantly different composition. Is this a problem for results? I guess this difference is way less strong than the age effect.

# Variance partitioning

Here I am trying to show the amount of variation in microbial structure explained by different explanatory variables or combinations of them.  
```{r variance partitioning, echo = FALSE}
# extract asv table
asvtab <- t(as(otu_table(XD.rare), "matrix"))

# variance partitioning structure
var <- varpart(asvtab, ~ age.days, ~ stage, ~ mass.g, ~ tank.id, data = meta.rare)
var2 <- varpart(asvtab, ~ age.days, ~ stage, ~ mass.g, ~ treatment, data = meta.rare)
var3 <- varpart(asvtab, ~ age.days, ~ stage, ~ body.condition, ~ treatment, data = meta.rare)
var4 <- varpart(asvtab, ~ age.days, ~ stage, ~ mass.g, ~ days.exposed, data = meta.rare)

# plots
plot(var, digits = 2, Xnames = c("Age", "Stage", "Mass", "Tank"), bg = c("yellow", "tomato", "lightblue", "lightgreen"))
plot(var2, digits = 2, Xnames = c("Age", "Stage", "Mass", "Treatment"), bg = c("yellow", "tomato", "lightblue", "lightgreen"))
plot(var3, digits = 2, Xnames = c("Age", "Stage", "Body Condition", "Treatment"), bg = c("yellow", "tomato", "lightblue", "lightgreen"))
plot(var4, digits = 2, Xnames = c("Age", "Stage", "Mass", "Days Exposed"), bg = c("yellow", "tomato", "lightblue", "lightgreen"))
```

Age clearly dominates in terms of its individual contribution, and stage seems to contribute mainly through its correlation with age. 

Treatment didn't have a strong individual effect which is good to see. Tank has a huge effect, which probably makes sense since those tadpoles will have the exact same bacterial exposures.

# (TO-DO) Differential abundance

*STILL NEED TO FIGURE OUT STRUCTURAL ZEROS*  
*STILL NEED TO CONFIRM LFC DIRECTION*

## Exposed vs not exposed

Now I want to see whether there is differential abundance of bacteria at the phylum level between tadpoles that have and haven't been exposed to thyroxine.  

Earlier I made a new Y/N thyroxine exposure metadata category based on number of days exposed.

```{r ANCOMBC exposure, warning = FALSE, message = FALSE, echo = FALSE}
set.seed(14)
ANCOM.exposure <- ancombc2(data = XD.complete,
                           tax_level = "Phylum",
                           fix_formula = "exposureYN",
                           p_adj_method = "BH",
                           group = "exposureYN",
                           struc_zero = TRUE,
                           verbose = TRUE)

tax.frame <- as.data.frame(tax_table(XD.complete))
tax.frame <- tax.frame %>% 
  mutate(TAXON = row.names(tax.frame))
res <- ANCOM.exposure$res %>% 
  filter(is.na(lfc_exposureYNYES) == FALSE)
df_fig1 <- data.frame(res$lfc_exposureYNYES * res$diff_exposureYNYES, check.names = FALSE) %>% 
  rownames_to_column("ASV")
colnames(df_fig1)[-1] = "BETA"
df_fig1 <- df_fig1 %>% 
  mutate(SD = res$se_exposureYNYES * res$diff_exposureYNYES) %>% 
  mutate(TAXON = res$taxon) %>% 
  filter(BETA != 0) %>% 
  arrange(BETA) %>%
  mutate(group = ifelse(BETA >0, "g1", "g2"))
df_fig.exposure = df_fig1 %>% left_join(tax.frame, by = "TAXON")

df_fig.exposure
# 
# zeros <- ANCOM.exposure$zero_ind
# # set up zeros comparison dataframe
# # remove taxa that were not present in either group or were present in both
# zeros_setup <- zeros %>% 
#   mutate(TAXON = zeros$taxon)
# ANCOM.all_zeros <- zeros_setup %>% left_join(tax.frame, by = "TAXON") %>% 
#   filter(!(`structural_zero (exposureYN = NO)` == TRUE & `structural_zero (exposureYN = YES)` == TRUE)) %>% 
#   filter(!(`structural_zero (exposureYN = NO)` == FALSE & `structural_zero (exposureYN = YES)` == FALSE)) %>% 
#   select(-taxon)
# ANCOM.all_zeros = zeros_setup %>% left_join(tax.frame, by = "TAXON")
```

## Pairwise by treatment
I can also compare all three groups with each other to see a more thorough breakdown.

```{r ANCOMBC treatment pairwise, warning = FALSE, message = FALSE, echo = FALSE}
set.seed(14)
ANCOM.treatment <- ancombc2(data = XD.complete,
                           tax_level = "Phylum",
                           fix_formula = "treatment",
                           p_adj_method = "BH",
                           group = "treatment",
                           struc_zero = TRUE,
                           verbose = TRUE,
                           pairwise = TRUE)

# Control vs Thyroxine
res <- ANCOM.treatment$res_pair %>% 
  filter(is.na(lfc_treatmentThyroxine) == FALSE)
df_fig1 <- data.frame(res$lfc_treatmentThyroxine * res$diff_treatmentThyroxine, check.names = FALSE) %>% 
  rownames_to_column("ASV")
colnames(df_fig1)[-1] = "BETA"
df_fig1 <- df_fig1 %>% 
  mutate(SD = res$se_treatmentThyroxine * res$diff_treatmentThyroxine) %>% 
  mutate(TAXON = res$taxon) %>% 
  filter(BETA != 0) %>% 
  arrange(BETA) %>%
  mutate(group = ifelse(BETA >0, "g1", "g2"))
df_fig.CT = df_fig1 %>% left_join(tax.frame, by = "TAXON")

# Control vs Delayed Thyroxine
res <- ANCOM.treatment$res_pair %>% 
  filter(is.na(`lfc_treatmentDelayed Thyroxine`) == FALSE)
df_fig1 <- data.frame(res$`lfc_treatmentDelayed Thyroxine` * res$`diff_treatmentDelayed Thyroxine`, check.names = FALSE) %>% 
  rownames_to_column("ASV")
colnames(df_fig1)[-1] = "BETA"
df_fig1 <- df_fig1 %>% 
  mutate(SD = res$`se_treatmentDelayed Thyroxine` * res$`diff_treatmentDelayed Thyroxine`) %>% 
  mutate(TAXON = res$taxon) %>% 
  filter(BETA != 0) %>% 
  arrange(BETA) %>%
  mutate(group = ifelse(BETA >0, "g1", "g2"))
df_fig.CDT = df_fig1 %>% left_join(tax.frame, by = "TAXON")

# Thyroxine vs Delayed Thyroxine
res <- ANCOM.treatment$res_pair %>% 
  filter(is.na(`lfc_treatmentThyroxine_treatmentDelayed Thyroxine`) == FALSE)
df_fig1 <- data.frame(res$`lfc_treatmentThyroxine_treatmentDelayed Thyroxine` * res$`diff_treatmentThyroxine_treatmentDelayed Thyroxine`, check.names = FALSE) %>% 
  rownames_to_column("ASV")
colnames(df_fig1)[-1] = "BETA"
df_fig1 <- df_fig1 %>% 
  mutate(SD = res$`se_treatmentThyroxine_treatmentDelayed Thyroxine` * res$`diff_treatmentThyroxine_treatmentDelayed Thyroxine`) %>% 
  mutate(TAXON = res$taxon) %>% 
  filter(BETA != 0) %>% 
  arrange(BETA) %>%
  mutate(group = ifelse(BETA >0, "g1", "g2"))
df_fig.TDT = df_fig1 %>% left_join(tax.frame, by = "TAXON")

print("Control vs Thyroxine")
df_fig.CT
print("Control vs Delayed Thyroxine")
df_fig.CDT
print("Thyroxine vs Delayed Thyroxine")
df_fig.TDT
```

