---
title: "XD Analysis"
author: "Jess Diaz"
date: "`r format(Sys.time(), '%a %d %b')`"
output:
  rmdformats::readthedown
---

# Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(qiime2R)
library(microbiome)
library(dplyr)
library(ggplot2)
library(vegan)
library(decontam)
library(phyloseq)
library(rmdformats)
source("HighstatLibV10.R")
library(ANCOMBC)
library(tibble)

theme_set(theme_light())
sc = scale_color_manual(values = c("black", "deepskyblue", "red"))
```

```{r functions, include=FALSE}
# calculates values for stats bars for plots
data.summary <- function(x) {
  m <- mean(x)
  ymin <- m-sd(x)
  ymax <- m+sd(x)
  return(c(y=m, ymin=ymin, ymax=ymax))
}
```

This is the analysis of the Xenopus tadpole thyroxine experiment. There are X samples plus 12 kit controls. These samples are spread across 3 groups: Thyroxine, Delayed Thyroxine, and Control.  

```{r create phyloseq object, include = FALSE}
# create master phyloseq
XD <-qza_to_phyloseq(
  features="table-filt-bytaxa.qza",
  tree="rooted-tree.qza",
  taxonomy="taxonomy.qza",
  metadata = "XD_qiime_metadata.txt"
)

# sum contrasts for categorical variables
sample_data(XD)$tank.id <- C(factor(sample_data(XD)$tank.id), sum)
sample_data(XD)$treatment <- C(factor(sample_data(XD)$treatment), sum)
```

# Data Cleanup

## Summary

```{r data summary, echo = FALSE}
# data summaries
summarize_phyloseq(XD)
```
## Decontam

First I will run decontam using the 12 kit control samples as the negative controls.

```{r decontam, echo = FALSE}
# make new variable where controls are "TRUE"
sample_data(XD)$is.neg <- sample_data(XD)$sample.type == "Blank"

# run decontam
contamdf.prevalence <- isContaminant(XD, method="prevalence", neg="is.neg", threshold=0.5)

# identify which taxa are contaminants
table(contamdf.prevalence$contaminant)
contams <- which(contamdf.prevalence$contaminant)

# list contaminants
tax <- as(tax_table(XD), "matrix")
tax[contams,]
```
I will also make a plot that shows prevalence in the negatives and in true samples, to check how they are getting classified.  
```{r decontam plot, echo = FALSE}
# presence-absence phyloseq for controls and samples
ps.pa <- transform_sample_counts(XD, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$sample.type == "Blank", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$sample.type == "Gut", ps.pa)

# make dataframe for prevalence
df.pa <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                    contaminant=contamdf.prevalence$contaminant)
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) +
  geom_point() +
  xlab("Prevalence (blanks)") +
  ylab("Prevalence (samples)") +
  ggtitle("Threshold = 0.5") +
  geom_abline(slope = 1, intercept = 0)
```

24 taxa were identified as contaminants using a threshold of 0.5. Based on the  plot above it seems reasonable to me to classify these as contaminants. Now I just want to make sure that I am not removing too many samples by removing these contaminants. This table shows % of reads removed from that sample after removing contaminants.  

```{r remove contaminants, echo = FALSE}
# remove taxa from phyloseq
XD.decontam <- prune_taxa(!contamdf.prevalence$contaminant, XD)

# set up df with readcounts and % removed
readcounts <- data.frame(matrix(ncol = 0, nrow = 291))
readcounts$pre <- as.data.frame(sample_sums(XD))
readcounts$post <- sample_sums(XD.decontam)
readcounts$per.removed <- ((readcounts$pre - readcounts$post)/readcounts$pre) *100

print(readcounts$per.removed)

# print otu table for contaminants
as.data.frame(otu_table(XD)) %>% slice(contams)
```
I am happy with how this looks, the couple samples with higher % of reads removed had low read counts anyway.

## Rarefaction

Now I can look at the read counts to see how many reads are in each sample.
```{r read counts, echo = FALSE}
sample_sums(XD.decontam)
```
Based on the read counts, I will rarefy at 3638 to retain all data points except the five that had fewer reads. Blanks should be automatically removed with the rarefaction.

```{r rarefy, include=FALSE}
XD.rare <- rarefy_even_depth(XD.decontam, sample.size=3638, rngseed=14, verbose=TRUE) # rarefy
```

I also need to remove the samples that were allowed to metamorphose.  

```{r remove metamorphose, include = FALSE}
# remove samples that were allowed to metamorphose
XD.rare <- subset_samples(XD.rare, age.days <= 100)
```

# Exploring Metadata

## Initial plots

```{r meta, echo = FALSE}
# isolate metadata for ggplot
meta.rare <- meta(XD.rare)
```

First I want to make the age/stage plot, colored by treatment group.
```{r Age vs Stage, echo = FALSE}
meta.rare %>% 
  ggplot(aes(x = age.days, y = stage, color = treatment)) +
  geom_point(size = 3, shape = 16) +
  stat_summary(geom = "line", fun = mean, linewidth = 1.5) +
  labs(x = "Age (days)", y = "Developmental Stage") +
  sc
```

Then I want to look at the relationship between body length and mass:

```{r body length and mass, echo = FALSE}
meta.rare %>% 
  ggplot(aes(x = length.mm, y = mass.g)) +
  geom_point(size = 3, aes(color = treatment)) + sc +
  labs(x = "Body length (mm)", y = "Mass (g)")
```
There don't seem to be any red points in the upper part of the graph, meaning the thyroxine group was smaller than the other groups. Maybe thyroxine stunted growth by accelerating development? Is is worth exploring this with a linear model?  

## Body Condition variable 

I want to make a new Body Condition measure, using the residuals of a linear regression between mass and body length (see Bancila et al 2010). First I want to transform the variables to give them a more linear relationship. Doing the log10 of both seems to work, and this is also what was done in Denoel et al 2002).  

```{r log body length and mass, echo = FALSE}
meta.rare %>% 
  ggplot(aes(x = log10(length.mm), y = log10(mass.g))) +
  geom_point(size = 3, aes(color = treatment)) + sc +
  labs(x = "Log10(Body length (mm))", y = "Log10(Mass (g))")
```

To find the index I will first make the log variables, then run the model and save the residuals as a body condition variable. (I am using this as an opportunity to also add the Y/N thyroxine exposure category to be used for differential abundance)  

```{r body condition model, echo = FALSE}
meta.rare <- meta.rare %>% 
  mutate(length.mm.log = log10(length.mm)) %>% 
  mutate(mass.g.log = log10(mass.g))

BCI.model <- lm(mass.g.log ~ length.mm.log, data = meta.rare)
summary(BCI.model)

plot(mass.g.log ~ length.mm.log, data = meta.rare)
x <- seq(0.38, 1.4, length.out = 10)
lines(x, -3.60608 + 2.82665 * x, col = "red")

meta.rare <- meta.rare %>% 
  mutate(body.condition = BCI.model$residuals) %>% 
  mutate(exposureYN = ifelse(days.exposed >0, "YES", "NO"))

# add body condition to phyloseq
body.table <- meta.rare %>%
  select(body.condition, exposureYN)
body.data <- sample_data(body.table)
XD.rare1 <- merge_phyloseq(XD.rare, body.data)
meta.rare1 <- meta(XD.rare1)
```

Now I can look at other variables in relation to body condition. For example, how does body condition change with age?  
```{r BCI and age, echo = FALSE}
meta.rare1 %>% 
  ggplot(aes(x = age.days, y = mass.g)) +
  geom_point(size = 3, aes(color = treatment)) + sc +
  labs(x = "Age (days)", y = "Mass (g)")

meta.rare1 %>% 
  ggplot(aes(x = age.days, y = body.condition)) +
  geom_point(size = 3, aes(color = treatment)) + sc +
  labs(x = "Age (days)", y = "Body condition")
```

Mass increases as the groups age but body condition doesn't, likely because they are growing proportionately in length as the get bigger in mass.

```{r BCI and treatment, echo = FALSE}
meta.rare1 %>% 
  ggplot(aes(x = treatment, y = body.condition)) +
  geom_boxplot(size = 1, aes(color = treatment)) + sc +
  labs(x = "Treatment", y = "Body Condition")
```
Treatment doesn't seem to impact body condition. What about the number of days exposed to thyroxine?  
```{r BCI and treatment, echo = FALSE}
meta.rare1 %>% 
  ggplot(aes(x = days.exposed, y = body.condition)) +
  geom_point(size = 3, aes(color = treatment)) + sc +
  labs(x = "Days exposed to thyroxine", y = "Body Condition")
```
Body condition also doesn't seem to be changing depending on how long the animals have been exposed to thyroxine.

## Statistics: Is the relationship between age and stage different depending on treatment? 

I can construct a pretty simple linear model using stage as the response variable and age and treatment as predictors. Add in the other variables too?

# Alpha diversity

## Plots

I can look at the initial alpha diversity plots to look for differences. These look pretty similar.  
```{r alpha all, echo = FALSE}
plot_richness(XD.rare1, x="treatment")
```
More zoomed in richness plots for each metadata variable:  
```{r alpha setup, include = FALSE}
# save alpha diversity calculations as alpha dataframe
alpha <- microbiome::alpha(XD.rare1)

# save phyloseq metadata as separate object, and add shannon and observed richness
meta.rare1 <- meta.rare1 %>%
  mutate(shannon = alpha$diversity_shannon) %>%
  mutate(observed = alpha$observed)
```

```{r alpha plots, echo = FALSE}
# by group
meta.rare1 %>%
  ggplot(aes(x=treatment, y=shannon)) + sc +
  geom_point(aes(color = treatment), position = position_dodge(width = 0.1), shape = 1, size = 3, alpha = 0.5) +
  stat_summary(fun.data = data.summary, geom = "errorbar", linewidth = 1, width = 0.5, na.rm = TRUE, aes(color = treatment)) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by age
meta.rare1 %>%
  ggplot(aes(x=age.days, y=shannon)) + sc +
  geom_point(aes(color = treatment), shape = 1, size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by stage
meta.rare1 %>%
  ggplot(aes(x=stage, y=shannon)) + sc +
  geom_point(aes(color = treatment), shape = 1, size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by days exposed
meta.rare1 %>%
  ggplot(aes(x=days.exposed, y=shannon)) + sc +
  geom_point(aes(color = treatment), shape = 1, size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by mass
meta.rare1 %>%
  ggplot(aes(x=mass.g, y=shannon)) + sc +
  geom_point(aes(color = treatment), shape = 1, size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")
```
No clear patterns here. Maybe diversity is increasing a little with age but overall there don't seem to be any visual relationships explaining variation in alpha diversity.  

## Statistics: Linear model

What factors drive alpha diversity variation? Can run a version of the biostats final project.  

# Beta diversity

## Plots

```{r beta div prep, include = FALSE}
BrayDist <- distance(XD.rare1, method = "bray")
BrayOrd <- ordinate(XD.rare1, "NMDS", distance = BrayDist)
BrayPlot <- plot_ordination(XD.rare1, BrayOrd, justDF = TRUE)
```

Plot of all samples:
```{r main beta div, echo = FALSE}
BrayPlot %>% 
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  geom_point(size = 3, aes(color = treatment)) + sc
```
Now I will color the plots by age and stage:
```{r colored bray curtis, echo = FALSE}
# colored by age
BrayPlot %>%
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  geom_point(size = 3, aes(color = age.days)) +
  scale_colour_gradient(low = "lightblue", high = "navy") +
  labs(color = "Age (days)")

# colored by stage
BrayPlot %>%
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  geom_point(size = 3, aes(color = stage)) +
  scale_colour_gradient(low = "lightblue", high = "navy") +
  labs(color = "Developmental Stage")
```

Age seems to have clearer separation over time compared to stage, but they both move from left to right. This likely explains the clustering of the Delayed Thyroxine group since they were all older tadpoles.  

## Statistics: PERMANOVA

Need to run this on the new data.  

```{r permanovas}
# MasterMetadata_no100_noDT <- MasterMetadata_no100 %>% 
#   filter(Treatment != "Delayed Thyroxine")
# 
# # by treatment - significant (might be due to age) (F = 6.8)
# adonis2(BrayDist ~ Treatment, data = MasterMetadata_no100)
# 
# # by age - significant F = 64.72)
# adonis2(BrayDist ~ Age_days, data = MasterMetadata_no100)
# 
# # by stage - significant (F = 27.998)
# adonis2(BrayDist ~ Stage, data = MasterMetadata_no100)
# 
# # by tank - significant (F = 1.8)
# adonis2(BrayDist ~ Tank_ID, data = MasterMetadata_no100)
# 
# adonis2(BrayDist ~ Age_days * Treatment, data = MasterMetadata_no100)
# 
# XEN_noDT <- XEN %>% 
#   subset_samples(Treatment != "Delayed Thyroxine")
# BrayDist_noDT <- distance(XEN_noDT, method = "bray")
# adonis2(BrayDist_noDT ~ Treatment, data = MasterMetadata_no100_noDT)
```

# Variance partitioning

Here I am trying to show the amount of variation in microbial structure explained by different explanatory variables or combinations of them.  
```{r variance partitioning, echo = FALSE}
# extract asv table
asvtab <- t(as(otu_table(XD.rare1), "matrix"))

# variance partitioning structure
var <- varpart(asvtab, ~ age.days, ~ stage, ~ mass.g, ~ tank.id, data = meta.rare)
var2 <- varpart(asvtab, ~ age.days, ~ stage, ~ mass.g, ~ treatment, data = meta.rare)
var3 <- varpart(asvtab, ~ age.days, ~ stage, ~ body.condition, ~ treatment, data = meta.rare)

# plots
plot(var, digits = 2, Xnames = c("Age", "Stage", "Mass", "Tank"), bg = c("yellow", "tomato", "lightblue", "lightgreen"))
plot(var2, digits = 2, Xnames = c("Age", "Stage", "Mass", "Treatment"), bg = c("yellow", "tomato", "lightblue", "lightgreen"))
plot(var3, digits = 2, Xnames = c("Age", "Stage", "Body Condition", "Treatment"), bg = c("yellow", "tomato", "lightblue", "lightgreen"))
```

Age clearly dominates in terms of its individual contribution, and stage seems to contribute mainly through its association with age.  
Treatment didn't have a strong individual effect which is good to see. Tank has a huge effect, which probably makes sense since those tadpoles will have the exact same bacterial exposures.

# Differential abundance

Now I want to see whether there is differential abundance of bacteria between tadpoles that have and haven't been exposed to thyroxine.  

Earlier I made a new Y/N thyroxine exposure metadata category based on number of days exposed.

```{r ANCOMBC exposure, warnings = FALSE, message = FALSE}
set.seed(14)
ANCOM.exposure <- ancombc2(data = XD.rare1,
                           tax_level = NULL,
                           fix_formula = "exposureYN",
                           p_adj_method = "BH",
                           group = "exposureYN",
                           verbose = TRUE)

tax.frame <- as.data.frame(tax_table(XD.rare1))
tax.frame <- tax.frame %>% 
  mutate(TAXON = row.names(tax.frame))
res <- ANCOM.exposure$res %>% 
  filter(is.na(lfc_exposureYNYES) == FALSE)
df_fig1 <- data.frame(res$lfc_exposureYNYES * res$diff_exposureYNYES, check.names = FALSE) %>% 
  rownames_to_column("ASV")
colnames(df_fig1)[-1] = "BETA"
df_fig1 <- df_fig1 %>% 
  mutate(SD = res$se_exposureYNYES * res$diff_exposureYNYES) %>% 
  mutate(TAXON = res$taxon) %>% 
  filter(BETA != 0) %>% 
  arrange(BETA) %>%
  mutate(group = ifelse(BETA >0, "g1", "g2"))
df_fig.exposure = df_fig1 %>% left_join(tax.frame, by = "TAXON")

df_fig.exposure
```

I can also compare all three groups with each other to see a more thorough breakdown.

```{r ANCOMBC treatment pairwise, warnings = FALSE, message = FALSE}
set.seed(14)
ANCOM.treatment <- ancombc2(data = XD.rare1,
                           tax_level = NULL,
                           fix_formula = "treatment",
                           p_adj_method = "BH",
                           group = "treatment",
                           verbose = TRUE,
                           pairwise = TRUE)

# Control vs Thyroxine
res <- ANCOM.treatment$res_pair %>% 
  filter(is.na(lfc_treatmentThyroxine) == FALSE)
df_fig1 <- data.frame(res$lfc_treatmentThyroxine * res$diff_treatmentThyroxine, check.names = FALSE) %>% 
  rownames_to_column("ASV")
colnames(df_fig1)[-1] = "BETA"
df_fig1 <- df_fig1 %>% 
  mutate(SD = res$se_treatmentThyroxine * res$diff_treatmentThyroxine) %>% 
  mutate(TAXON = res$taxon) %>% 
  filter(BETA != 0) %>% 
  arrange(BETA) %>%
  mutate(group = ifelse(BETA >0, "g1", "g2"))
df_fig.CT = df_fig1 %>% left_join(tax.frame, by = "TAXON")

# Control vs Delayed Thyroxine
res <- ANCOM.treatment$res_pair %>% 
  filter(is.na(`lfc_treatmentDelayed Thyroxine`) == FALSE)
df_fig1 <- data.frame(res$`lfc_treatmentDelayed Thyroxine` * res$`diff_treatmentDelayed Thyroxine`, check.names = FALSE) %>% 
  rownames_to_column("ASV")
colnames(df_fig1)[-1] = "BETA"
df_fig1 <- df_fig1 %>% 
  mutate(SD = res$`se_treatmentDelayed Thyroxine` * res$`diff_treatmentDelayed Thyroxine`) %>% 
  mutate(TAXON = res$taxon) %>% 
  filter(BETA != 0) %>% 
  arrange(BETA) %>%
  mutate(group = ifelse(BETA >0, "g1", "g2"))
df_fig.CDT = df_fig1 %>% left_join(tax.frame, by = "TAXON")

# Thyroxine vs Delayed Thyroxine
res <- ANCOM.treatment$res_pair %>% 
  filter(is.na(`lfc_treatmentThyroxine_treatmentDelayed Thyroxine`) == FALSE)
df_fig1 <- data.frame(res$`lfc_treatmentThyroxine_treatmentDelayed Thyroxine` * res$`diff_treatmentThyroxine_treatmentDelayed Thyroxine`, check.names = FALSE) %>% 
  rownames_to_column("ASV")
colnames(df_fig1)[-1] = "BETA"
df_fig1 <- df_fig1 %>% 
  mutate(SD = res$`se_treatmentThyroxine_treatmentDelayed Thyroxine` * res$`diff_treatmentThyroxine_treatmentDelayed Thyroxine`) %>% 
  mutate(TAXON = res$taxon) %>% 
  filter(BETA != 0) %>% 
  arrange(BETA) %>%
  mutate(group = ifelse(BETA >0, "g1", "g2"))
df_fig.TDT = df_fig1 %>% left_join(tax.frame, by = "TAXON")
```
