---
title: "XD Analysis"
author: "Jess Diaz"
date: "2023-05-09"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: false
    theme: cosmo
    highlight: monochrome
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(qiime2R)
library(microbiome)
library(dplyr)
library(ggplot2)
library(vegan)
library(decontam)
library(phyloseq)
source("HighstatLibV10.R")

theme_set(theme_light())
c = scale_color_manual(values = c("black", "deepskyblue", "red"))
```

## Setup

```{r create phyloseq object, include = FALSE}
# create master phyloseq
XD <-qza_to_phyloseq(
  features="table-filt-bytaxa.qza",
  tree="rooted-tree.qza",
  taxonomy="taxonomy.qza",
  metadata = "XD_qiime_metadata.txt"
)

# sum contrasts for categorical variables
sample_data(XD)$tank.id <- C(factor(sample_data(XD)$tank.id), sum)
sample_data(XD)$treatment <- C(factor(sample_data(XD)$treatment), sum)
```

## Data Cleanup and Exploration

```{r data summary, echo = FALSE}
# data summaries
summarize_phyloseq(XD)
```
First I will run decontam using the 12 kit control samples as the negative controls.

```{r decontam, echo = FALSE}
# make new variable where controls are "TRUE"
sample_data(XD)$is.neg <- sample_data(XD)$sample.type == "Blank"

# run decontam
contamdf.prevalence <- isContaminant(XD, method="prevalence", neg="is.neg", threshold=0.5)

# identify which taxa are contaminants
table(contamdf.prevalence$contaminant)
contams <- which(contamdf.prevalence$contaminant)

# list contaminants
tax <- as(tax_table(XD), "matrix")
tax[contams,]

# remove taxa from phyloseq
XD.decontam <- prune_taxa(!contamdf.prevalence$contaminant, XD)
```

24 taxa were identified as contaminants using a threshold where the taxa have to be more prevalent in the negatives than in the samples. I have removed these taxa from analysis.

Now I can look at the read counts to see how many reads are in each sample.
```{r}
sample_sums(XD.decontam)
```
Based on the read counts, I will rarefy at 3638 to retain all data points except the five that had fewer reads. Blanks should be automatically removed with the rarefaction.

```{r rarefy, include=FALSE}
XD.rare <- rarefy_even_depth(XD.decontam, sample.size=3638, rngseed=14, verbose=TRUE) # rarefy
```
I also need to remove the samples that were allowed to metamorphose.
```{r}
# remove samples that were allowed to metamorphose
XD.rare <- subset_samples(XD.rare, age.days <= 100)
```

## Exploring Metadata

```{r}
# isolate metadata for ggplot
meta.rare <- meta(XD.rare)
```

First I want to make the age/stage plot, colored by treatment group.
```{r Age vs Stage, echo = FALSE}
meta.rare %>% 
  ggplot(aes(x = age.days, y = stage, color = treatment)) +
  geom_point(size = 3, shape = 16) +
  stat_summary(geom = "line", fun = mean, linewidth = 1.5) +
  labs(x = "Age (days)", y = "Developmental Stage") +
  c
```

Then I want to look at the relationship between body length and mass:
```{r}
meta.rare %>% 
  ggplot(aes(x = length.mm, y = mass.g)) +
  geom_point(size = 3, aes(color = treatment)) + c +
  labs(x = "Body length (mm)", y = "Mass (g)")
```
There don't seem to be any red points in the upper part of the graph, meaning the thyroxine group was smaller than the other groups. Maybe thyroxine stunted growth by accelerating development?
### Statistics: Is the relationship between age and stage different depending on treatment
Linear model here

NEXT STEPS:  

* Create new BMI variable, check out the papers Kevin sent
* Run linear model for age vs stage

## Alpha diversity
## Beta diversity
## Variance partitioning
