---
title: "XD Analysis"
author: "Jess Diaz"
date: "`r format(Sys.time(), '%a %d %b')`"
output:
  rmdformats::readthedown
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(qiime2R)
library(microbiome)
library(dplyr)
library(ggplot2)
library(vegan)
library(decontam)
library(phyloseq)
library(rmdformats)
source("HighstatLibV10.R")

theme_set(theme_light())
sc = scale_color_manual(values = c("black", "deepskyblue", "red"))
```

```{r functions, include=FALSE}
# calculates values for stats bars for plots
data.summary <- function(x) {
  m <- mean(x)
  ymin <- m-sd(x)
  ymax <- m+sd(x)
  return(c(y=m, ymin=ymin, ymax=ymax))
}
```

This is the analysis of the Xenopus tadpole thyroxine experiment. There are X samples plus 12 kit controls. These samples are spread across 3 groups: Thyroxine, Delayed Thyroxine, and Control.

## Setup

```{r create phyloseq object, include = FALSE}
# create master phyloseq
XD <-qza_to_phyloseq(
  features="table-filt-bytaxa.qza",
  tree="rooted-tree.qza",
  taxonomy="taxonomy.qza",
  metadata = "XD_qiime_metadata.txt"
)

# sum contrasts for categorical variables
sample_data(XD)$tank.id <- C(factor(sample_data(XD)$tank.id), sum)
sample_data(XD)$treatment <- C(factor(sample_data(XD)$treatment), sum)
```

## Data Cleanup and Setup

```{r data summary, echo = FALSE}
# data summaries
summarize_phyloseq(XD)
```
### Decontam

First I will run decontam using the 12 kit control samples as the negative controls.

```{r decontam, echo = FALSE}
# make new variable where controls are "TRUE"
sample_data(XD)$is.neg <- sample_data(XD)$sample.type == "Blank"

# run decontam
contamdf.prevalence <- isContaminant(XD, method="prevalence", neg="is.neg", threshold=0.5)

# identify which taxa are contaminants
table(contamdf.prevalence$contaminant)
contams <- which(contamdf.prevalence$contaminant)

# list contaminants
tax <- as(tax_table(XD), "matrix")
tax[contams,]
```
I will also make a plot that shows prevalence in the negatives and in true samples, to check how they are getting classified.  
```{r decontam plot, echo = FALSE}
# presence-absence phyloseq for controls and samples
ps.pa <- transform_sample_counts(XD, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$sample.type == "Blank", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$sample.type == "Gut", ps.pa)

# make dataframe for prevalence
df.pa <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                    contaminant=contamdf.prevalence$contaminant)
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) +
  geom_point() +
  xlab("Prevalence (blanks)") +
  ylab("Prevalence (samples)") +
  ggtitle("Threshold = 0.5") +
  geom_abline(slope = 1, intercept = 0)
```

24 taxa were identified as contaminants using a threshold of 0.5. Based on the  plot above it seems reasonable to me to classify these as contaminants. Now I just want to make sure that I am not removing too many samples by removing these contaminants. This table shows % of reads removed from that sample after removing contaminants.  
```{r remove contaminants, echo = FALSE}
# remove taxa from phyloseq
XD.decontam <- prune_taxa(!contamdf.prevalence$contaminant, XD)

# set up df with readcounts and % removed
readcounts <- data.frame(matrix(ncol = 0, nrow = 291))
readcounts$pre <- as.data.frame(sample_sums(XD))
readcounts$post <- sample_sums(XD.decontam)
readcounts$per.removed <- ((readcounts$pre - readcounts$post)/readcounts$pre) *100

print(readcounts$per.removed)

# print otu table for contaminants
as.data.frame(otu_table(XD)) %>% slice(contams)
```
I am happy with how this looks, the couple samples with higher % of reads removed had low read counts anyway.

### Rarefaction

Now I can look at the read counts to see how many reads are in each sample.
```{r read counts, echo = FALSE}
sample_sums(XD.decontam)
```
Based on the read counts, I will rarefy at 3638 to retain all data points except the five that had fewer reads. Blanks should be automatically removed with the rarefaction.

```{r rarefy, include=FALSE}
XD.rare <- rarefy_even_depth(XD.decontam, sample.size=3638, rngseed=14, verbose=TRUE) # rarefy
```

I also need to remove the samples that were allowed to metamorphose.  
```{r remove metamorphose, include = FALSE}
# remove samples that were allowed to metamorphose
XD.rare <- subset_samples(XD.rare, age.days <= 100)
```

## Exploring Metadata

```{r meta, echo = FALSE}
# isolate metadata for ggplot
meta.rare <- meta(XD.rare)
```

First I want to make the age/stage plot, colored by treatment group.
```{r Age vs Stage, echo = FALSE}
meta.rare %>% 
  ggplot(aes(x = age.days, y = stage, color = treatment)) +
  geom_point(size = 3, shape = 16) +
  stat_summary(geom = "line", fun = mean, linewidth = 1.5) +
  labs(x = "Age (days)", y = "Developmental Stage") +
  sc
```

Then I want to look at the relationship between body length and mass:
```{r body length and mass, echo = FALSE}
meta.rare %>% 
  ggplot(aes(x = length.mm, y = mass.g)) +
  geom_point(size = 3, aes(color = treatment)) + sc +
  labs(x = "Body length (mm)", y = "Mass (g)")
```
There don't seem to be any red points in the upper part of the graph, meaning the thyroxine group was smaller than the other groups. Maybe thyroxine stunted growth by accelerating development?  

### Statistics: Is the relationship between age and stage different depending on treatment?  

Linear model here  

NEXT STEPS:  

* Create new BMI variable, check out the papers Kevin sent
* Run linear model for age vs stage

## Alpha diversity
I can look at the initial alpha diversity plots to look for differences. These look pretty similar.  
```{r alpha all, echo = FALSE}
plot_richness(XD.rare, x="treatment")
```
More zoomed in richness plots for each metadata variable:  
```{r alpha setup, include = FALSE}
# save alpha diversity calculations as alpha dataframe
alpha <- microbiome::alpha(XD.rare)

# save phyloseq metadata as separate object, and add shannon and observed richness
meta.rare <- meta(XD.rare) %>%
  mutate(shannon = alpha$diversity_shannon) %>%
  mutate(observed = alpha$observed)
```

```{r alpha plots, echo = FALSE}
# by group
meta.rare %>%
  ggplot(aes(x=treatment, y=shannon)) + sc +
  geom_point(aes(color = treatment), position = position_dodge(width = 0.1), shape = 1, size = 3, alpha = 0.5) +
  stat_summary(fun.data = data.summary, geom = "errorbar", linewidth = 1, width = 0.5, na.rm = TRUE, aes(color = treatment)) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by age
meta.rare %>%
  ggplot(aes(x=age.days, y=shannon)) + sc +
  geom_point(aes(color = treatment), shape = 1, size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by stage
meta.rare %>%
  ggplot(aes(x=stage, y=shannon)) + sc +
  geom_point(aes(color = treatment), shape = 1, size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by age
meta.rare %>%
  ggplot(aes(x=age.days, y=shannon)) + sc +
  geom_point(aes(color = treatment), shape = 1, size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by days exposed
meta.rare %>%
  ggplot(aes(x=days.exposed, y=shannon)) + sc +
  geom_point(aes(color = treatment), shape = 1, size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")

# by mass
meta.rare %>%
  ggplot(aes(x=mass.g, y=shannon)) + sc +
  geom_point(aes(color = treatment), shape = 1, size = 3) +
  theme_classic(base_size = 15) +
  ylab("shannon diversity")
```

### Statistics: Linear model

## Beta diversity
```{r beta div prep, include = FALSE}
BrayDist <- distance(XD.rare, method = "bray")
BrayOrd <- ordinate(XD.rare, "NMDS", distance = BrayDist)
BrayPlot <- plot_ordination(XD.rare, BrayOrd, justDF = TRUE)
```

Plot of all samples:
```{r main beta div, echo = FALSE}
BrayPlot %>% 
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  geom_point(size = 3, aes(color = treatment)) + sc
```
Now I will color the plots by age and stage:
```{r colored bray curtis, echo = FALSE}
# colored by age
BrayPlot %>%
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  geom_point(size = 3, aes(color = age.days)) +
  scale_colour_gradient(low = "lightblue", high = "navy") +
  labs(color = "Age (days)")

# colored by stage
BrayPlot %>%
  ggplot(aes(x = NMDS1, y = NMDS2)) +
  geom_point(size = 3, aes(color = stage)) +
  scale_colour_gradient(low = "lightblue", high = "navy") +
  labs(color = "Developmental Stage")
```


Age seems to have clearer separation over time compared to stage, but they both move from left to right.\

### Statistics: PERMANOVA
```{r permanovas}
# MasterMetadata_no100_noDT <- MasterMetadata_no100 %>% 
#   filter(Treatment != "Delayed Thyroxine")
# 
# # by treatment - significant (might be due to age) (F = 6.8)
# adonis2(BrayDist ~ Treatment, data = MasterMetadata_no100)
# 
# # by age - significant F = 64.72)
# adonis2(BrayDist ~ Age_days, data = MasterMetadata_no100)
# 
# # by stage - significant (F = 27.998)
# adonis2(BrayDist ~ Stage, data = MasterMetadata_no100)
# 
# # by tank - significant (F = 1.8)
# adonis2(BrayDist ~ Tank_ID, data = MasterMetadata_no100)
# 
# adonis2(BrayDist ~ Age_days * Treatment, data = MasterMetadata_no100)
# 
# XEN_noDT <- XEN %>% 
#   subset_samples(Treatment != "Delayed Thyroxine")
# BrayDist_noDT <- distance(XEN_noDT, method = "bray")
# adonis2(BrayDist_noDT ~ Treatment, data = MasterMetadata_no100_noDT)
```

## Variance partitioning

Here I am trying to show the amount of variation in microbial structure explained by different explanatory variables or combinations of them.  
```{r variance partitioning, echo = FALSE}
# extract asv table
asvtab <- t(as(otu_table(XD.rare), "matrix"))

# variance partitioning structure
var <- varpart(asvtab, ~ age.days, ~ stage, ~ mass.g, ~ tank.id, data = meta.rare)
var2 <- varpart(asvtab, ~ age.days, ~ stage, ~ mass.g, ~ treatment, data = meta.rare)

# plots
plot(var, digits = 2, Xnames = c("Age", "Stage", "Mass", "Tank"), bg = c("yellow", "tomato", "lightblue", "lightgreen"))
plot(var2, digits = 2, Xnames = c("Age", "Stage", "Mass", "Treatment"), bg = c("yellow", "tomato", "lightblue", "lightgreen"))
```